File: hackerrank/cracking_the_coding_interview/data_structures/linked_lists_detect_cycle/detect_cycle.rb
# Head pointer may be nil if the list is empty
# Return 0 for no cycle
# Return 1 for cycle exists in linked list
# Use a hash to track nodes that have been visited
# Traverse linked list until next pointer is nil

class Node
  attr_accessor :data, :next

  def initialize(data = nil, next_node = nil)
    @data = data
    @next = next_node
  end
end

def has_cycle(head)
  return 0 unless head

  visited_nodes = Hash.new

  current_node = head
  while current_node.next
    return 1 if visited_nodes[current_node]
    visited_nodes[current_node] = true
    current_node = current_node.next
  end

  return 0
end

File: hackerrank/cracking_the_coding_interview/data_structures/linked_lists_detect_cycle/detect_cycle.py
def has_cycle(head):
  if head == None:
    return 1
  
  visited_nodes = {}

  current_node = head

  while (current_node.next != None):
    if visited_nodes.get(current_node) == None:
      return 1

    visited_nodes[current_node] = True

    current_node = current_node.next
  
  return 0
